[lldb/crashlog] Fix module binary resolution

medismailben opened the pull request:
This patch fixes a bug in when resolving and loading modules from the binary image list.

When loading a module, we would first use the UUID from the binary image list with `dsymForUUID` to fetch the dSYM bundle from our remote build records and copy the executable locally.

If we failed to find a matching dSYM bundle for that UUID on the build record, let's say if that module was built locally, we use Spotlight (`mdfind`) to find the dSYM bundle once again using the UUID.

Prior to this patch, we would set the image path to be the same as the symbol file. This resulted in trying to load the dSYM as a module in lldb, which isn't allowed.

This patch address that by looking for a binary matching the image identifier, next to the dSYM bundle and try to load that instead.

rdar://127433616

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -418,9 +418,22 @@ def locate_module_and_debug_symbols(self):
+                        for binary in find_results:
+                            abs_path = os.path.abspath(binary)
+                            basename = os.path.basename(binary)
+                            if os.path.exists(abs_path) and basename == self.identifier:
```
Do we expect `abs_path` to not exist after the call to `find`?

JDevlieghere commented on lldb/examples/python/crashlog.py:
```diff
@@ -418,9 +418,22 @@ def locate_module_and_debug_symbols(self):
+                        parent_dir = os.path.dirname(dsym)
+                        find_results = (
+                            subprocess.check_output(
+                                '/usr/bin/find "%s" -type f \( -perm -u=x -o -perm -g=x -o -perm -o=x \)'
```
Rather than shelling out to find, can you use `os.walk`?

bulbazord commented on lldb/examples/python/crashlog.py:
```diff
@@ -418,9 +418,20 @@ def locate_module_and_debug_symbols(self):
+                        for root, dirs, files in os.walk(parent_dir):
+                            for file in files:
+                                abs_path = os.path.join(root, file)
+                                if os.path.isfile(abs_path) and os.access(
```
This must necessarily be a file right? It's in the `files` output of `os.walk`. Could it be something else?

medismailben replied:
It could be block device or another file type fwiw.

JDevlieghere commented on lldb/examples/python/crashlog.py:
```diff
@@ -418,9 +418,20 @@ def locate_module_and_debug_symbols(self):
+                        # Look for the executable next to the dSYM bundle.
+                        parent_dir = os.path.dirname(dsym)
+                        executables = []
+                        for root, dirs, files in os.walk(parent_dir):
```
Let's use `_` instead of `dirs` as it's unused.

medismailben committed:
[lldb/crashlog] Fix module binary resolution

This patch fixes a bug in when resolving and loading modules from the
binary image list.

When loading a module, we would first use the UUID from the binary image
list with `dsymForUUID` to fetch the dSYM bundle from our remote build
records and copy the executable locally.

If we failed to find a matching dSYM bundle for that UUID on the build
record, let's say if that module was built locally, we use Spotlight
(`mdfind`) to find the dSYM bundle once again using the UUID.

Prior to this patch, we would set the image path to be the same as the
symbol file. This resulted in trying to load the dSYM as a module in
lldb, which isn't allowed.

This patch address that by looking for a binary matching the image
identifier, next to the dSYM bundle and try to load that instead.

rdar://127433616

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
